### DISCUSSION TITLE
JMX Metrics Enabled on Zookeeper/Datadog

### DISCUSSION THREAD

**Discussion by tferfce** [strimzi/strimzi-kafka-operator#9404] [Q&A] [answered]
Hello,

So I'm trying to use jmx metrics that are exposed from zookeeper and collect them with zk integration for datadog. I have also configured the the openmetrics integration with prometheus exporter which is supported from Kafka crd and it works like charm. When i try to use this though it doesn't collect and zookeeper specific metrics.

Works:
{
  "openmetrics": {
    "init_config": {
      "service": "Zookeeper"
    },
    "instances": [
      {
        "openmetrics_endpoint": "http://%%host%%:9404/metrics",
        "namespace": "zookeeper",
        "metrics": [
          ".*"
        ]
      }
    ]
  }
}
Doesn't Work:
{
  "zk": {
    "init_config": {
      "is_jmx": true,
      "collect_default_metrics": true
    },
    "instances": [
      {
        "host": "%%host%%",
        "port": "2181",
        "timeout": 120,
        "name": "JMX Zookeeper"
      }
    ]
  }
}

The only metrics im able to collect are these. 
zookeeper.instances
zookeeper.instances.inactive

My question is which metrics are exposed from this configuration??

![zookeeper_config](https://github.com/strimzi/strimzi-kafka-operator/assets/68991289/dbe770b8-1250-47f1-9cda-79008a8c6581)

Thanks,
Thomas

**Comment by scholzj** [comment]
The JMX access definitely won't be on port 2181 where you seem to point in your JSON but on port 9999.

**Reply by scholzj** [reply]
Also, JMX does not expose some metrics. It opens a port where clients can connect with the JMX client and talk with the server.  So it is up to your metrics server what it does with the JMX interface, what metrics it connects, whether it uses it to configure things or shutdown things etc. (as JMX is not only a monitoring but a management solution).

**Comment by tferfce** [comment]
I already tried the 9999 port and i get this error 
![image](https://github.com/strimzi/strimzi-kafka-operator/assets/68991289/ea45a2e6-da27-433a-9f40-5aa934038035)
And just to note that this happens only with zk integration. With jmx integration where i jvm metrics are collected have no issues! On port 2181 i wasn't getting these errors.

**Reply by scholzj** [reply]
Wait, hold on, I'm confused. You talk about too many things ... `zk integration`, `jmx integration`. Strimzi supports getting information about ZooKeeper in two ways:
* As Prometheus metrics on port 9404
* Use the JMX interface on port 9999

So, what do you mean with `zk integration`? The screenshot does not make it obvious what it is for application, what it is doing, nothing. So it does not help much. If you are actually trying to connect to ZooKeeper with the ZooKeeper protocol, that will not work as port 2181 is secured using mTLS and it is not supported for other applications to connect to it. So you would need to hack around that.

**Reply by tferfce** [reply]
Datadog supports zk integration which is documented in here [ZK Integration](https://docs.datadoghq.com/integrations/zk/?tab=host). The jmx integration is a different one that supports the port you mentioned and its actually working for zookeeper [JMX Integration](https://docs.datadoghq.com/integrations/java/?tab=host). The reason i was trying to connect on 2181 is the zk integration. As for the mTLS you might be right cause cause it is mentioned to use SSL and to configure the DD agent to use SSL to connect to zookeeper. Prometheus metrics with openmetrics [Openmetrics Integration](https://docs.datadoghq.com/integrations/openmetrics/) is working for 9404.

Prometheus: 9094 works
JMX: 9999 works on jmx but not on zk (error)
ZK: 8121 doesn't collect

**Reply by scholzj** [reply]
Ok, well, as I said, Strimzi does not support any third party applications connecting directly to ZooKeeper. So I do not think that will work for you.

**Reply by tferfce** [reply]
Do you think that enabling the SSL communication with DD and Zookeeper might have something to do with it and worth the try or the Prometheus metrics is the only way in my case?

**Reply by scholzj** [reply, marked as answer]
I think the Prometheus metrics are all that is really needed. Also, keep in mind that you are not missing _enabling TLS communication_. As I said, this is not supported in Strimzi. You would need to _hack_ mTLS security - e.g. _steal_ and reuse the certificates of the Kafka brokers or similar component in order to pass the authentication. And if you decide to do it, you should properly understand what will DataDog do with the ZooKeeper cluster as it can do basically anything it wants with it as it is not just a metrics export.

**Reply by tferfce** [reply]
Hmm i understand. It carries a security risk on that! Very much appreciate the support! Thanks!
